### ISSUE TITLE
Minimum system memory requirements

### ISSUE THREAD

**Issue by escoand** [matiasdelellis/facerecognition#319]
Tried to test this app on [Odoid-HC2](https://www.hardkernel.com/shop/odroid-hc2-home-cloud-two/) but it's failing with `Your system does not meet the minimum of memory requirements`. This system has 2GB of RAM but this is slightly less than the expected 2GB. Anyway, is this requirement still correct when using model 3 which takes round about 100MB of memory? Setup the app with model 3 works flawlessly in contrast to model 1.

```
$ grep MemTotal /proc/meminfo 
MemTotal:        2043852 kB
$ echo $((2043852 * 1024))
2092904448
$ echo $((2 * 1024 * 1024 * 1024))
2147483648
```

**Comment by matiasdelellis** [comment]
HI @escoand 

> Anyway, is this requirement still correct when using model 3 which takes round about 100MB of memory? 

This is regarding the analysis of the images, but the face clustering could also consume a lot of memory depending on the number of images. Therefore, beyond the model, there is a general application requirement. We wanted to be prudent using 2GB, but maybe changed it to 1Gb.

Meanwhile, try to change this limit by changing this line.

https://github.com/matiasdelellis/facerecognition/blob/master/lib/Service/SettingsService.php#L38

**Comment by escoand** [comment]
@matiasdelellis Thanks for the instant answer. Did something similar and it's working.

**Comment by escoand** [comment]
One off-topic question: My father uses Google Picasa since years on old hardware and the face detection is working quite well and fast. How do they do this without this requirements?

**Comment by matiasdelellis** [comment]
> One off-topic question: My father uses Google Picasa since years on old hardware and the face detection is working quite well and fast. How do they do this without this requirements?

haha. I don't know ...
It would be nice to research it. :sweat_smile: 

As I recall, it wasn't magic either. It was still an intensive task, but it certainly worked with any decent hardware.
Keep in mind that probably your dad in all this time has just a couple of thousand photos. Now with phones of 32Gb/64Gb/128GB, we sure have much more.. In the last survey I did, there were users who reported over 100000 photos. I don't think picasa scales well with so many images.

Also for this reason, the application is designed as a service that works in the background. No matter how long it takes, (Although of course we try to optimize it). The important thing is to obtain the best results, working little by little, and see the results progressively.

On the other hand, I am sure they only found absolutely frontal faces, and in any configuration we find more faces. Try model 1, with 640x480 images and you will surely have better results at a decent speed.. :wink:

**Comment by cratoo** [comment]
"In the last survey I did, there were users who reported over 100000 photos. I don't think picasa scales well with so many images."

I'm closely following the development of this as I hope to replace my picasa stuff with nextcloud tooling. And I'm one of these people with multiple 10k's of photos and picasa scales well with it. The only problem there exists is that the UX becomes a problem with such an amount of photos. Of course you hardly do face recognition of multiple 10k's of photos again and again.
As far as I remember my research showed that picasa uses sqlite in the background. Most likely the reason for the performance.

**Comment by escoand** [comment]
@matiasdelellis This brings me to the next question. I really like my Odroid for being small and still capable enough for Nextcloud. But what is the best way to scan the images on more decent hardware and import it afterwards into my running Nextcloud?

@cratoo But there is maybe a migration path from Picasa to this. Aren't the face coordinates saved in the `picasa.ini` file? Or anywhere else we could transform them into XMP information and someday this app will be able to import this information.

**Comment by cratoo** [comment]
@escoand yes they're. There is even some documentation online, how the coordinates are saved (https://gist.github.com/stigkj/2392516). Picasa even was able to write XMP as separate files until one of the last versions released.

**Comment by matiasdelellis** [comment]
> @matiasdelellis This brings me to the next question. I really like my Odroid for being small and still capable enough for Nextcloud. But what is the best way to scan the images on more decent hardware and import it afterwards into my running Nextcloud?
> @cratoo But there is maybe a migration path from Picasa to this. Aren't the face coordinates saved in the picasa.ini file? Or anywhere else we could transform them into XMP information and someday this app will be able to import this information.

There is still no way. There are requests to enable external models, to anilize with rented GPUs, etc. but this breaks the entire concept of privacy of the application. 

I'd like to add some option to rely on XMP tags, but first nextcloud infrastructure is missing to get it right.

Note it probably wouldn't work as well as analyzing them from scratch here since the 'recognizer' depends a lot on the rectangle used, and the proposed rectangle cannot be reliable for our application.

The same apply to picassa.ini.

However, I don't think it's likely to import from applications files. This would open the Pandora's box for the application.  If I import from picasa, I would have to do it from digikam, and any other application, which will honestly cause more problems than solutions. Sorry.

So, I hope at some point to trust XMP regions, but it is highly unlikely that I will consider importing from applications.

**Comment by escoand** [comment]
@matiasdelellis Sure, always use public interfaces. And XMP is well defined. I don't want to import from Picasa directly.

**Comment by escoand** [comment]
I've tested [Facebox](https://machinebox.io/docs/facebox) by MachineBox. The detection is quite limited with just frontal faces. But they have a nice HTTP api. You are able to upload an image and get in response a JSON object with the coordinates and names of the faces. This is a simple approach to do the hard work on a more powerful machine. It's a separate Docker container on a local machine, so still no privacy issue.

**Comment by matiasdelellis** [comment]
>I've tested Facebox by MachineBox.

Great. 
I did not know him.

> The detection is quite limited with just frontal faces.

Try model 3 with small images and you will get a very similar result. It's fast, and you will get the most important faces.

> But they have a nice HTTP api. You are able to upload an image and get in response a JSON object with the coordinates and names of the faces.

We will not incorporate any model that needs previous training. It has to be generic, send an image, and it has to return the rectangle and the descriptor to compare.. The name is not part of the model.!

> This is a simple approach to do the hard work on a more powerful machine. It's a separate Docker container on a local machine, so still no privacy issue.

Please, even if it is docker, don't say anything about privacy.. :disappointed: :sweat_smile: 
You can believe that it is safe, (Although it is very rare that they asks for registration to instantiate the docker.), but think it as a service. Your users trust to upload data to your nextcloud instance, believing that their data does not leave your server. What if you tell them that you send all their private photos to a 3rd party service? :thinking:

**Comment by matiasdelellis** [comment]
Fixed on #326
Thank fol all.

**Comment by escoand** [comment]
> We will not incorporate any model that needs previous training. It has to be generic, send an image, and it has to return the rectangle and the descriptor to compare.. The name is not part of the model.!

Facebox is working in the same way. For test purpose I've created a small script to upload an image. All faces are unknown but with the coordinates I cropped the faces and saved them as separated files. Second step, I've labeled this faces and uploaded to the teach endpoint. Now I could upload more pictures and they are recognized.
This is a simple workflow. Yes, not for a normal user but It's quite easy to work with when you are able to handle a shell.

> Please, even if it is docker, don't say anything about privacy.. :disappointed: :sweat_smile: 
> You can believe that it is safe, (Although it is very rare that they asks for registration to instantiate the docker.), but think it as a service. Your users trust to upload data to your nextcloud instance, believing that their data does not leave your server. What if you tell them that you send all their private photos to a 3rd party service? :thinking: 

In my case it's just me and my family and it's also my personal interest to have everything local.

Thanks for the change...
